src/monster.cpp

```diff
@@ -162,6 +163,36 @@ void monster::poly(mtype *t)
     faction = t->default_faction;
 }
 
+void monster::update_check(){
+    if (type->upgrade_group == "NULL"){
+        return;
+    }
+    int current_day = calendar::turn.get_turn()/ DAYS(1);
+    int upgrade_time = type->upgrade_min * ACTIVE_WORLD_OPTIONS["MONSTER_GROUP_DIFFICULTY"];
+    //g->u.add_msg_if_player(m_debug, "Upgrade group: %s ", type->upgrade_group);
+    add_msg(m_debug, "Current:day: %d", current_day);
+    add_msg(m_debug, "Upgrade time : %d", upgrade_time);
+    add_msg(m_debug, "Last loaded: %d", last_loaded);
+
+    if (current_day == last_loaded || last_loaded < upgrade_time){
+        add_msg(m_debug, "Upgrade time less");
+        last_loaded = current_day;
+        return;
+    }
+
+    int time_passed = current_day - last_loaded;
+    add_msg(m_debug, "Time passed: %d", time_passed);
+    //radioactive decay function
+    float upgrade_chance =100 * (1- pow(0.5 , (float(time_passed) / float(type->half_life )))) + type->base_upgrade_chance;
+    add_msg(m_debug, "Upgrade chance: %f", upgrade_chance);
+    if (upgrade_chance > rng(0, 100)){
```

Comment:
Also as best as I can tell this should be rng(0,99). Else wise we will always have a basic 1/101 chance of not upgrading, no matter how many half lives we might have gone through.

Another smallish thing that I'm noting as a side effect of just doing 0-99 is that the minimum success rate is capped at 1% per time loaded (since .5^X when X!= 0 will always be > 0). This means that for any monster with a half life of > 70 days is gonna start seeing some weirdness happen based on how many times they are loaded. A rather easy way to fix this at no real cost would just be to have upgrade_chance be 1000 * X and then do rng(0,999), which would mean that the weirdness would only happen with half lives > 700 days and would be much less likely to happen IMO. (Maybe include a comment denoting 700 days as the max half life that the algorithm should handle due to insufficient accuracy, with a note to switch to a double + larger rng range should we need a longer half life).